src/tests/state.rs

```diff
@@ -359,7 +359,16 @@ impl RepositoryClient for TestRepositoryClient {
     }
 
     async fn set_branch_to_sha(&mut self, branch: &str, sha: &CommitSha) -> anyhow::Result<()> {
-        self.add_branch_sha(branch, &sha.0);
+        let branch_history = self.branch_history.entry(branch.to_string()).or_default();
```

Comment:
I don't think that this logic is needed, we should know how many resets have happened. Let's remove this logic and allow duplicates there (and modify the test so that it checks that the same SHA was really set twice).